### PR TITLE
 Add unit tests for integration cli utils.go functions

### PR DESCRIPTION
**Pull request by vdemeester**
This adds tests for a part of integration-cli utilities code 🐙 : `utils.go` and `checker.go` (almost tempted to but [this image](http://www.daedtech.com/wp-content/uploads/2012/12/TestAllTheThings.jpg) :smile_cat:).

This is a start of the work discussed a bit in #15121 :

> - tests to ensure error handling works as expected
> - tests to check that the functions do what they're meant to do
> - test the functions with the testing framework to ensure that they behave the way we assume they behave in the context of integration-cli

As the "integration-cli utilities code" is starting to be quite big (more than 2000 lines of code), it can feel safer/better to have test that covers these codes (just like testing framework have their own unit tests). 

Back to the PR/code :
- This moves `waitRun`, `waitContainer` and `waitInspect` to `docker_utils.go` file as these 3 functions are the only ones in there that are directly related to the docker binary (uses the `dockerBinary`). It felt more natural to put them in `docker_utils.go`.
- This moves _a part of_ the integration-cli utils code to `pkg/integration` package (to make it unit-testable easily and even shared possibly :smile:) ; but keeps `utils.go` for now (to make the transition less painful — will act as wrappers kinda).
- This adds some comments on each methods about what they are supposed to do (just to make it more clear how or which one to use it when writing integration-cli tests).
- This adds a bunch of unit tests on each function (some might be poorly written — if so, let me know :stuck_out_tongue_closed_eyes:).
- This adds some simple test for `integration-cli/checker` too :smiley_cat: (and moved it to `pkg/integration/checker`).

What probably needs to be improved :
- Some tests and the timing on some of them :sweat_smile: 
- <del>The way the tests are run. For now they are run when `test-integration-cli` is and this looks like that</del>. Unit tests for integration are now run with `test-unit` :

``` bash
# […]
+ go test -test.timeout=60m github.com/docker/docker/pkg/integration
PASS
coverage: 91.9% of statements

+ go test -test.timeout=60m github.com/docker/docker/pkg/integration/checker
OK: 2 passed
PASS
coverage: 90.9% of statements
# […]
```

The next part will be to test `docker_utils.go` and the other files ; probably split these files (mostly `docker_utils.go` which is ~1500k lines of code) in smaller one that groups coherent function / struct (like the `Daemon` sturcture).

🐸

/cc @unclejack @tiborvass @duglin @cpuguy83 @LK4D4 

Signed-off-by: Vincent Demeester vincent@sbr.pm

### CODE REVIEW

**Comment by vdemeester** [comment]
```bash
---> Making bundle: validate-test (in bundles/1.9.0-dev/validate-test)
These files use the wrong testing infrastructure:
 - integration-cli/utils_test.go
```

Well… I didn't see it coming :sweat_smile: :sweat: :scream_cat:

File: hack/make/validate-test [hunk 1]
```diff
@@ -11,7 +11,7 @@ unset IFS
 badFiles=()
 for f in "${files[@]}"; do
 	# skip check_test.go since it *does* use the testing package
-	if [ "$f" = "integration-cli/check_test.go" ]; then
+	if [ "$f" = "integration-cli/check_test.go" ] || [ "$f" = "integration-cli/utils_test.go" ]; then
```
Comment (vdemeester) [inline hack/make/validate-test]: This is temporary (and as is pretty bad) to make sure the validation does not "mis-interpret" my unit tests with integration tests. Will have to find a cleaner solution though.

File: integration-cli/utils_test.go [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+		t.Fatal(err)
+	}
+
+	if string(b) != expected {
+		t.Fatalf("Expected '%s', got '%s'", expected, string(b))
+	}
+}
+
+// FIXME doesn't work
+// func TestRunAtDifferentDate(t *testing.T) {
```
Comment (vdemeester) [inline integration-cli/utils_test.go]: Commented this out as I didn't find a way to make it work :sob:. Note that this function is only used for Notary integration test that are, as of now, skipped with `c.Skip("Currently changes system time, causing instability")`.

**Comment by jfrazelle** [comment]
this is so meta <3

File: hack/make/validate-test (hunk 1 above)
Comment (tiborvass) [inline hack/make/validate-test]: wow yeah that sucks :S Any ideas on the solution?

Comment (vdemeester) [inline hack/make/validate-test]: Right now no :sweat:, but I'll try to find the best way :smile:

**Comment by vdemeester** [comment]
Ok I need to adapt the tests to work on windows and linux/unix too (using different commands for the tests..).

File: hack/make/validate-test [hunk 3]
```diff
@@ -5,16 +5,15 @@
 source "${MAKEDIR}/.validate"
 
 IFS=$'\n'
-files=( $(validate_diff --diff-filter=ACMR --name-only -- 'integration-cli/*.go' || true) )
+
+# skip few files :
+# - check_test.go since it *does* use the testing package
+# - checker and utils folder since it is the "test" framework and using testing to be unit-tested
+files=( $(validate_diff --diff-filter=ACMR --name-only -- 'integration-cli/*.go' | grep -vE "^integration-cli/utils|^integration-cli/checker|^integration-cli/check_test.go" || true) )
```
Comment (vdemeester) [inline hack/make/validate-test]: Not sure it's better from previous one but.. it's a little bit cleaner

File: hack/make/test-integration-cli [hunk 4]
```diff
@@ -2,6 +2,8 @@
 set -e
 
 bundle_test_integration_cli() {
+	go_test_dir ./integration-cli/checker
+	go_test_dir ./integration-cli/utils
```
Comment (vdemeester) [inline hack/make/test-integration-cli]: This runs the unit testss before the integration-cli ones.

File: integration-cli/utils/utils_test.go [hunk 5]
```diff
@@ -0,0 +1,490 @@
+// +build !windows
```
Comment (vdemeester) [inline integration-cli/utils/utils_test.go]: I don't like this too much as it's gonna be skipped on Windows but.. :sweat_smile:

**Comment by calavera** [comment]
:space_invader: LGTM :space_invader:

**Comment by jfrazelle** [comment]
LGTM

**Comment by vdemeester** [comment]
:tada:

Now it's gonna be the fun part, `docker_utils.go` :smile_cat:

**Pull request merged**